### ISSUE TITLE
"RuntimeError: can't modify frozen Hash" when using with rack-cors

### ISSUE THREAD

**Issue by teejayvanslyke** [hirefire/hirefire-resource#33]
Hello:

Today we upgraded our application to use hirefire-resource 0.3.8. We're also using the [rack-cors](https://github.com/cyu/rack-cors) middleware. It appears that when we make an info request to the HireFire endpoint, rack-cors raises:

```
RuntimeError: can't modify frozen Hash

…undle/ruby/2.0.0/gems/rack-cors-0.4.0/lib/rack/
cors.rb:  96:in `[]='
…undle/ruby/2.0.0/gems/rack-cors-0.4.0/lib/rack/
cors.rb:  96:in `call'
…by/2.0.0/gems/unicorn-4.9.0/lib/unicorn/
http_server.rb: 580:in `process_client'
…by/2.0.0/gems/unicorn-4.9.0/lib/unicorn/
http_server.rb: 674:in `worker_loop'
…by/2.0.0/gems/unicorn-4.9.0/lib/unicorn/
http_server.rb: 529:in `spawn_missing_workers'
…by/2.0.0/gems/unicorn-4.9.0/lib/unicorn/
http_server.rb: 140:in `start'
```

We're using Rails 4.2.4. Locking hirefire-resource at the version we were previously using (0.1.1) fixes the problem.

Please let me know if you need any more information!

**Comment by meskyanichi** [comment]
Thanks for reporting this issue. Does this issue occur both in development as well as in production?

**Comment by meskyanichi** [comment]
I've just generated a new Rails application to see if I can reproduce the issue, but I'm currently unable to.

These are the steps I took:

Generated a new Rails 4.2.4 application, and added the following gems:

```rb
gem "rack-cors", "0.4.0", require: "rack/cors"
gem "hirefire-resource", "0.3.8"
```

Then I added a basic rack-cors configuration to `config/application.rb`:

```rb
require File.expand_path("../boot", __FILE__)
require "rails/all"
Bundler.require(*Rails.groups)

module Issue33
  class Application < Rails::Application
    config.middleware.insert_before 0, "Rack::Cors" do
      allow do
        origins "*"
        resource "*",
          headers: :any,
          methods: [:get, :post, :options]
      end
    end
  end
end
```

Then I added `config/initializers/hirefire.rb`:

```rb
HireFire::Resource.configure do |config|
  config.dyno(:worker) do
    rand(100)
  end
end
```

Both middlewares are injected:

```
$ rake middleware
use HireFire::Middleware
use Rack::Cors
...
```

When booting the server with `rails s` I visit `http://localhost:3000/hirefire/development/info` and I see:

```json
[{"name":"worker","quantity":16}]
```

I don't see anything regarding the frozen hash issue.

**Comment by meskyanichi** [comment]
If possible, could you upgrade `hirefire-resource` one version at a time to see at which version the issue first occurs? That'll probably help narrow down the cause.

**Comment by blakecash** [comment]
@meskyanichi I was having the same issue.. My middleware was inserting Rack::Cors before HireFire::Middleware.

Here is my gem setup and I am on rails 4.2

```
gem "rack-cors", "0.4.0", require: "rack/cors"
gem "hirefire-resource", "0.3.8"
```

```
$ rake middleware                                                                                                        
use Rack::Cors
use HireFire::Middleware
...
```

I added this to the bottom of my middleware.rb initializer file.

```
  config.middleware.delete        'HireFire::Middleware'
  config.middleware.insert        'Rack::Cors', 'HireFire::Middleware'
```

Now everything is working and the middleware is in the correct order

```
$ rake middleware
use HireFire::Middleware                                                                                                        
use Rack::Cors
...
```

**Comment by meskyanichi** [comment]
@blakecash thanks for reporting that! I'm not sure why `Rack::Cors` gets injected at the top of the stack in your application, while in my blank application it gets injected after `HireFire::Middleware`.

How are you configuring and injecting `Rack::Cors`?

**Comment by mrrooijen** [comment]
[This commit](https://github.com/hirefire/hirefire-resource/commit/539f1d995953ab1f9b7aed0973f20fa520c84cfb) introduced a bug and is causing the issue. I've patched and released v0.3.11 which fixes it. The workaround by @blakecash https://github.com/hirefire/hirefire-resource/issues/33#issuecomment-157288548 should no longer be necessary.
